core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -95,199 +155,679 @@ public static RecordState forId(byte id) {
      */
     private int maxDeliveryCount;
 
-    private long nextFetchOffset = 0;
+    /**
+     * The share partition end offset specifies the partition end offset from which the records
+     * are already fetched.
+     */
+    private long endOffset;
 
-    SharePartition(int maxInFlightMessages, int maxDeliveryCount) {
+    /**
+     * The next fetch offset specifies the partition offset which should be fetched from the leader
+     * for the share partition.
+     */
+    private long nextFetchOffset;
+
+    SharePartition(String groupId, TopicIdPartition topicIdPartition, int maxInFlightMessages, int maxDeliveryCount) {
+        this.groupId = groupId;
+        this.topicIdPartition = topicIdPartition;
         this.maxInFlightMessages = maxInFlightMessages;
         this.maxDeliveryCount = maxDeliveryCount;
         this.cachedState = new ConcurrentSkipListMap<>();
+        this.lock = new ReentrantReadWriteLock();
+        // Should be initialized when the partition is loaded by reading state from the share persister.
+        this.nextFetchOffset = 0;
+        this.endOffset = 0;
+        this.findNextFetchOffset = new AtomicBoolean(false);
         // TODO: Just a placeholder for now.
         assert this.maxInFlightMessages > 0;
         assert this.maxDeliveryCount > 0;
     }
 
     public long nextFetchOffset() {
-        return nextFetchOffset;
+        if (findNextFetchOffset.compareAndSet(true, false)) {
+            lock.writeLock().lock();
+            try {
+                // Find the next fetch offset.
+                long baseOffset = nextFetchOffset;
+                Map.Entry<Long, InFlightBatch> floorEntry = cachedState.floorEntry(nextFetchOffset);
+                if (floorEntry != null) {
+                    baseOffset = floorEntry.getKey();
+                }
+                // The next fetch offset is the next offset after the last fetched batch.
+                NavigableMap<Long, InFlightBatch> subMap = cachedState.subMap(baseOffset, true, endOffset, true);
+                if (subMap.isEmpty()) {
+                    nextFetchOffset = endOffset + 1;
+                } else {
+                    boolean updated = false;
+                    for (Map.Entry<Long, InFlightBatch> entry : subMap.entrySet()) {
+                        if (entry.getValue().offsetState() == null) {
```

Comment:
I am not sure in which situation is this value `null`. Can we add a comment for that?

Reply:
Added.